### DISCUSSION TITLE
Validate a JWT token

### DISCUSSION THREAD

**Discussion by dimaip** [zitadel/zitadel#5263] [Q&A]
Hi!
Sorry, a very n00b question.

I have a SPA app to which I authorise with Zitadel with PKCE.
Now I want to be able to call my server-side api as an authorised user.

After logging in with Zitadel, I receive a JWT token and an authorisation token with which I can call the Zitadel api.
Now, if I pass the authorisation token to my api, I'd be able to call Zitadel's API `/oauth/v2/introspect` and verify that the user is actually who he claims to be. But that would imply an extra request, which ideally I'd like to avoid.

I'd rather use a JWT token and assert user's identity based on it, but I don't know where to get the JWT secret to be able to validate it. Is there a way to pass this secret to Zitadel as a setting or something?
I couldn't find it.

**Comment by dakshitha** [comment]
It's a great question! You can change the token type in the token settings in your application to JWT from Bearer Token. 

<img width="1350" alt="Screenshot 2023-02-22 at 20 28 46" src="https://user-images.githubusercontent.com/11885043/220662437-080a8f53-9f4a-40af-9024-dd21ebf89eae.png">

The downside of not calling the introspection endpoint, however, is that the user might have logged out before the expiration of the token and you won't be able to know this without calling the introspection endpoint. You could call the introspection endpoint and cache the response for a set period of time without calling the endpoint all the time. Regardless of whether the token is of type JWT or opaque, you can do the introspection. 

Hope this helps :)

**Reply by dimaip** [reply]
Thanks for your reply, @dakshitha!
Please pardon me, but I still don't understand how I can validate that token server-side, without calling Zitadel.
In order to validate a JWT token, I need to know with which secret that token has been generated, and as I understand that secret is stored within Zitadel's DB, and it's not possible to configure it manually.

Caching Zitadel calls could be an option, but I was wondering if there's a simpler way.

> The downside of not calling the introspection endpoint, however, is that the user might have logged out before the expiration

That's totally fine for my usecase.

**Reply by dimaip** [reply]
And yes, in general, what's the point of providing the user with an identity JWT token, if there's no way to validate it outside Zitadel? Why not use a bearer token in all case?

**Reply by fforootd** [reply]
Well the id_token is specified on the OIDC Spec. and MUST be a JWT. The access token generally should be treated as opaque. But since the rise of JWT many people want to use JWTs as well for access_tokens.

To verify either the id_token or the access_token when they are in JWT format can be done on the server by fetching the public keys that where used to sign the token. In zitadel you can find the keys endpoint when looking for the "jwks_uri" (default is /oauth/v2/keys) in the discovery endpoint (default is /.well-known/openid-configuration)

hope this helps

**Reply by dimaip** [reply]
> by fetching the public keys that where used to sign the token

Ohhh that sure does, exactly what I've been asking for! Thanks so much!

> the id_token is specified on the OIDC Spec

Thanks for the pointer, I'll make sure to read into the spec to have more awareness of what's going on.
